File: sdk/python/examples/mdap/src/mdap/mdap.py
"""
MDAP Orchestration for FlatAgent v0.6.0.

Implements the first-to-ahead-by-k voting mechanism from the MAKER paper.
Handles multi-sampling, regex parsing, validation, and state management.
"""

import copy
import json
import re
from collections import Counter
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional, Tuple

import jsonschema
import litellm

from flatagents import FlatAgent, setup_logging, get_logger

# Configure logging
setup_logging(level='INFO')
logger = get_logger(__name__)


@dataclass
class MDAPConfig:
    """MDAP-specific configuration."""
    k_margin: int = 3
    max_candidates: int = 10
    max_steps: int = 100
    max_response_tokens: int = 2048


@dataclass
class MDAPMetrics:
    """Execution metrics collected during MDAP runs."""
    total_samples: int = 0
    total_red_flags: int = 0
    red_flags_by_reason: Dict[str, int] = field(default_factory=dict)
    samples_per_step: List[int] = field(default_factory=list)

    def record_red_flag(self, reason: str):
        self.total_red_flags += 1
        self.red_flags_by_reason[reason] = self.red_flags_by_reason.get(reason, 0) + 1


class MDAPOrchestrator:
    """
    MDAP voting orchestrator for FlatAgent v0.6.0.

    Uses the agent for prompt rendering, handles LLM calls, regex parsing,
    validation, and voting. State management is external to the agent.
    """

    def __init__(
        self,
        agent: FlatAgent,
        config: Optional[MDAPConfig] = None
    ):
        self.agent = agent
        self.config = config or self._load_config_from_agent()
        self.metrics = MDAPMetrics()

        # Load parsing and validation from metadata
        self._parsing_config = agent.metadata.get('parsing', {})
        self._validation_schema = agent.metadata.get('validation', None)

        # Compile regex patterns
        self._patterns = {}
        for field_name, field_config in self._parsing_config.items():
            pattern = field_config.get('pattern')
            if pattern:
                self._patterns[field_name] = (
                    re.compile(pattern, re.DOTALL),
                    field_config.get('type', 'str')
                )

    def _load_config_from_agent(self) -> MDAPConfig:
        """Load MDAP config from agent's metadata."""
        mdap_config = self.agent.metadata.get('mdap', {})
        return MDAPConfig(
            k_margin=mdap_config.get('k_margin', 3),
            max_candidates=mdap_config.get('max_candidates', 10),
            max_steps=mdap_config.get('max_steps', 100),
            max_response_tokens=mdap_config.get('max_response_tokens', 2048),
        )

    def _parse_response(self, content: str) -> Optional[Dict[str, Any]]:
        """Parse LLM response using regex patterns from config."""
        if not self._patterns:
            return None

        result = {}
        for field_name, (pattern, field_type) in self._patterns.items():
            match = pattern.search(content)
            if match:
                value = match.group(1)
                if field_type == 'json':
                    try:
                        result[field_name] = json.loads(value)
                    except json.JSONDecodeError:
                        return None
                elif field_type == 'int':
                    try:
                        result[field_name] = int(value)
                    except ValueError:
                        return None
                else:
                    result[field_name] = value
            else:
                return None  # Required field missing

        return result

    def _validate_parsed(self, parsed: Dict[str, Any]) -> bool:
        """Validate parsed result against JSON Schema from config."""
        if not self._validation_schema:
            return True

        try:
            jsonschema.validate(instance=parsed, schema=self._validation_schema)
            return True
        except jsonschema.ValidationError:
            return False

    async def sample_once(self, input_data: Dict[str, Any]) -> Tuple[str, Optional[Dict[str, Any]]]:
        """
        Make one LLM call and parse the response.

        Args:
            input_data: Input for the agent (available as {{ input.* }})

        Returns:
            Tuple of (raw_content, parsed_result or None)
        """
        # Render prompts using agent's templates
        system_prompt = self.agent._render_system_prompt(input_data)
        user_prompt = self.agent._render_user_prompt(input_data)

        messages = [
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": user_prompt}
        ]

        # Make LLM call
        response = await litellm.acompletion(
            model=self.agent.model,
            messages=messages,
            temperature=self.agent.temperature,
            max_tokens=self.agent.max_tokens,
        )

        # Track usage
        self.agent.total_api_calls += 1
        if hasattr(response, 'usage') and response.usage:
            input_tokens = getattr(response.usage, 'prompt_tokens', 0)
            output_tokens = getattr(response.usage, 'completion_tokens', 0)
            self.agent.total_cost += (input_tokens * 0.001 + output_tokens * 0.002) / 1000

        content = response.choices[0].message.content

        # Handle null content
        if content is None:
            return "", None

        # Parse response
        parsed = self._parse_response(content)

        return content, parsed

    def _check_red_flags(self, content: str, parsed: Optional[Dict[str, Any]]) -> Optional[str]:
        """Check response for red flags per MAKER paper."""
        # Format flag: parsing failed
        if parsed is None:
            return "format_error"

        # Validation flag
        if not self._validate_parsed(parsed):
            return "validation_failed"

        # Length flag: response too long
        estimated_tokens = len(content) // 4
        if estimated_tokens > self.config.max_response_tokens:
            return "length_exceeded"

        return None

    async def first_to_ahead_by_k(
        self,
        input_data: Dict[str, Any]
    ) -> Tuple[Optional[Dict[str, Any]], int]:
        """
        Core voting logic - sample until one response leads by k_margin.

        Returns:
            Tuple of (winning_response, num_samples)
        """
        votes: Counter = Counter()
        responses: Dict[str, Dict[str, Any]] = {}
        num_samples = 0

        for _ in range(self.config.max_candidates):
            try:
                content, parsed = await self.sample_once(input_data)
                num_samples += 1
                self.metrics.total_samples += 1

                # Red-flag check
                flag_reason = self._check_red_flags(content, parsed)
                if flag_reason:
                    self.metrics.record_red_flag(flag_reason)
                    logger.debug(f"Red-flagged: {flag_reason}")
                    continue

                logger.debug(f"Parsed result: {parsed}")

                # Vote on parsed result
                key = json.dumps(parsed, sort_keys=True)
                votes[key] += 1
                responses[key] = parsed

                # Check if single response hits k_margin
                if votes[key] >= self.config.k_margin:
                    logger.debug(f"Winner found with {votes[key]} votes after {num_samples} samples")
                    return parsed, num_samples

                # Check if leader ahead by k_margin
                if len(votes) >= 2:
                    top = votes.most_common(2)
                    if top[0][1] - top[1][1] >= self.config.k_margin:
                        logger.debug(f"Leader ahead by k after {num_samples} samples")
                        return responses[top[0][0]], num_samples

            except Exception as e:
                logger.warning(f"Sample failed: {e}")
                continue

        # Majority fallback
        if votes:
            winner_key = votes.most_common(1)[0][0]
            logger.debug(f"Majority fallback after {num_samples} samples")
            return responses[winner_key], num_samples

        logger.warning("No valid responses obtained")
        return None, num_samples

    @property
    def total_api_calls(self) -> int:
        return self.agent.total_api_calls

    @property
    def total_cost(self) -> float:
        return self.agent.total_cost


def create_orchestrator_from_config(config_path: str) -> MDAPOrchestrator:
    """Create orchestrator from YAML config file."""
    agent = FlatAgent(config_file=config_path)
    return MDAPOrchestrator(agent)

File: sdk/examples/mdap/js/src/mdap/mdap.ts
import { FlatAgent } from 'flatagents';
import Ajv, { ValidateFunction } from 'ajv';

type ParsingConfig = Record<string, { pattern: string; type?: string }>;

type MDAPConfig = {
  k_margin: number;
  max_candidates: number;
  max_steps: number;
  max_response_tokens: number;
};

type MDAPMetrics = {
  total_samples: number;
  total_red_flags: number;
  red_flags_by_reason: Record<string, number>;
  samples_per_step: number[];
};

function stableStringify(value: unknown): string {
  if (value === null || typeof value !== 'object') {
    return JSON.stringify(value);
  }
  if (Array.isArray(value)) {
    return `[${value.map(stableStringify).join(',')}]`;
  }
  const obj = value as Record<string, unknown>;
  const keys = Object.keys(obj).sort();
  const entries = keys.map(key => `${JSON.stringify(key)}:${stableStringify(obj[key])}`);
  return `{${entries.join(',')}}`;
}

export class MDAPOrchestrator {
  private agent: FlatAgent;
  config: MDAPConfig;
  metrics: MDAPMetrics;
  private parsingConfig: ParsingConfig;
  private validationSchema?: Record<string, any>;
  private patterns: Record<string, { regex: RegExp; type: string }> = {};
  private validator?: ValidateFunction;

  constructor(agent: FlatAgent, config?: Partial<MDAPConfig>) {
    this.agent = agent;

    const metadata = (agent as any).config?.metadata ?? {};
    const mdapConfig = metadata.mdap ?? {};

    this.config = {
      k_margin: config?.k_margin ?? mdapConfig.k_margin ?? 3,
      max_candidates: config?.max_candidates ?? mdapConfig.max_candidates ?? 10,
      max_steps: config?.max_steps ?? mdapConfig.max_steps ?? 100,
      max_response_tokens: config?.max_response_tokens ?? mdapConfig.max_response_tokens ?? 2048,
    };

    this.metrics = {
      total_samples: 0,
      total_red_flags: 0,
      red_flags_by_reason: {},
      samples_per_step: [],
    };

    this.parsingConfig = metadata.parsing ?? {};
    this.validationSchema = metadata.validation ?? undefined;

    for (const [fieldName, fieldConfig] of Object.entries(this.parsingConfig)) {
      if (fieldConfig.pattern) {
        this.patterns[fieldName] = {
          regex: new RegExp(fieldConfig.pattern, 's'),
          type: fieldConfig.type ?? 'str',
        };
      }
    }

    if (this.validationSchema) {
      const ajv = new Ajv({ allErrors: true });
      this.validator = ajv.compile(this.validationSchema);
    }
  }

  private parseResponse(content: string): Record<string, any> | null {
    if (!Object.keys(this.patterns).length) {
      return null;
    }

    const result: Record<string, any> = {};
    for (const [fieldName, { regex, type }] of Object.entries(this.patterns)) {
      const match = regex.exec(content);
      if (!match) {
        return null;
      }
      const value = match[1];
      if (type === 'json') {
        try {
          result[fieldName] = JSON.parse(value);
        } catch {
          return null;
        }
      } else if (type === 'int') {
        const parsed = Number.parseInt(value, 10);
        if (Number.isNaN(parsed)) {
          return null;
        }
        result[fieldName] = parsed;
      } else {
        result[fieldName] = value;
      }
    }

    return result;
  }

  private validateParsed(parsed: Record<string, any> | null): boolean {
    if (!parsed) return false;
    if (!this.validator) return true;
    return Boolean(this.validator(parsed));
  }

  private recordRedFlag(reason: string): void {
    this.metrics.total_red_flags += 1;
    this.metrics.red_flags_by_reason[reason] = (this.metrics.red_flags_by_reason[reason] ?? 0) + 1;
  }

  private checkRedFlags(content: string, parsed: Record<string, any> | null): string | null {
    if (!parsed) {
      return 'format_error';
    }
    if (!this.validateParsed(parsed)) {
      return 'validation_failed';
    }
    const estimatedTokens = Math.floor(content.length / 4);
    if (estimatedTokens > this.config.max_response_tokens) {
      return 'length_exceeded';
    }
    return null;
  }

  async sampleOnce(inputData: Record<string, any>): Promise<{ content: string; parsed: Record<string, any> | null }> {
    const result = await this.agent.call(inputData);
    const content = result.content ?? '';
    const parsed = this.parseResponse(content);
    return { content, parsed };
  }

  async firstToAheadByK(inputData: Record<string, any>): Promise<{ result: Record<string, any> | null; samples: number }> {
    const votes = new Map<string, number>();
    const responses = new Map<string, Record<string, any>>();
    let numSamples = 0;

    for (let i = 0; i < this.config.max_candidates; i += 1) {
      try {
        const { content, parsed } = await this.sampleOnce(inputData);
        numSamples += 1;
        this.metrics.total_samples += 1;

        const flagReason = this.checkRedFlags(content, parsed);
        if (flagReason) {
          this.recordRedFlag(flagReason);
          continue;
        }

        if (!parsed) {
          continue;
        }

        const key = stableStringify(parsed);
        votes.set(key, (votes.get(key) ?? 0) + 1);
        responses.set(key, parsed);

        const count = votes.get(key) ?? 0;
        if (count >= this.config.k_margin) {
          return { result: parsed, samples: numSamples };
        }

        if (votes.size >= 2) {
          const sorted = Array.from(votes.entries()).sort((a, b) => b[1] - a[1]);
          if (sorted[0][1] - sorted[1][1] >= this.config.k_margin) {
            const winner = responses.get(sorted[0][0]) ?? null;
            return { result: winner, samples: numSamples };
          }
        }
      } catch {
        continue;
      }
    }

    if (votes.size > 0) {
      const winner = Array.from(votes.entries()).sort((a, b) => b[1] - a[1])[0][0];
      return { result: responses.get(winner) ?? null, samples: numSamples };
    }

    return { result: null, samples: numSamples };
  }
}
